### PR TITLE
feat(medusa): Modules initializer

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
What:
Module initializer for loading modules into projects.

Example:

``` typescript
import { InventoryServiceInitializeOptions, initialize } from "@medusajs/inventory"

const options: InventoryServiceInitializeOptions = {
  database: {
    type: "postgres",
    url: DB_URL,
  },
}

const inventoryService = await initialize(options)
const newInventoryItem = await inventoryService.createInventoryItem({
  sku: "sku_123",
})

console.log(newInventoryItem)
```

Fixes: CORE-1140

### CODE REVIEW

File: packages/modules-sdk/src/medusa-module.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  MODULE_SCOPE,
+} from "./types"
+import { loadModuleMigrations } from "./loaders/utils"
+
+const logger: any = {
+  log: (a) => console.log(a),
+  info: (a) => console.log(a),
+  warn: (a) => console.warn(a),
+  error: (a) => console.error(a),
+}
```
Comment (carlos-r-l-rodrigues) [inline packages/modules-sdk/src/medusa-module.ts:17]: This will be addresed in another PR and Logger will be moved to `medusa-core-utils`

File: packages/inventory/package.json [hunk 2]
```diff
@@ -17,25 +17,20 @@
...
     "cross-env": "^5.2.1",
     "jest": "^25.5.4",
     "ts-jest": "^25.5.1",
     "typescript": "^4.4.4"
   },
   "dependencies": {
-    "@medusajs/modules-sdk": "*",
-    "awilix": "^8.0.0",
-    "typeorm": "^0.3.11"
+    "@medusajs/medusa": "workspace:^",
```
Comment (adrien2p) [inline packages/inventory/package.json:26]: question: Is that temporary? When the package will be deployed, not sure it works? Also, I suppose this deps is temporary until we extract the necessary elements right?

File: packages/inventory/src/migrations/run-migration.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { InternalModuleDeclaration, LoaderOptions } from "@medusajs/modules-sdk"
+import { DataSource, DataSourceOptions } from "typeorm"
+import { InventoryServiceInitializeOptions } from "../types"
+
+import migrations from "./index"
+
+function getDataSource(
+  dbData: InventoryServiceInitializeOptions["database"]
+): DataSource {
+  return new DataSource({
```
Comment (adrien2p) [inline packages/inventory/src/migrations/run-migration.ts:10]: suggestion: Like in the core databaseLoader, export from the file a `const datasource` as it must be globally available if you want to add repositories in the future. Wdyt? That way if someone else try to add a repo, he will not have to think about it :)

File: packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts [hunk 4]
```diff
@@ -96,7 +94,7 @@ export const createVariantTransaction = async (
       return
     }
 
-    return await inventoryServiceTx!.createInventoryItem({
+    return await inventoryService!.createInventoryItem({
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts:97]: suggestion: if we put `!` on the service i means that we are sure it will be present. In that case I would remove the `?` on the type

File: packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts [hunk 5]
```diff
@@ -111,7 +109,7 @@ export const createVariantTransaction = async (
 
   async function removeInventoryItem(inventoryItem: InventoryItemDTO) {
     if (inventoryItem) {
-      await inventoryServiceTx!.deleteInventoryItem(inventoryItem.id)
+      await inventoryService!.deleteInventoryItem(inventoryItem.id)
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts:112]: suggestion: Same as previous

File: packages/medusa/src/commands/utils/get-migrations.js [hunk 6]
```diff
@@ -252,3 +250,43 @@ export const getModuleSharedResources = (configModule, featureFlagsRouter) => {
     migrations,
   }
 }
+
+export const runIsolatedModulesMigration = async (configModule) => {
```
Comment (adrien2p) [inline packages/medusa/src/commands/utils/get-migrations.js:254]: 🙏 very clean

File: packages/modules-sdk/src/medusa-module.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    const moduleResolutions = registerMedusaModule(moduleKey, modDeclaration!)
+
+    await moduleLoader({ container, moduleResolutions, logger })
+
+    const services = {}
+
+    for (const resolution of Object.values(moduleResolutions)) {
+      const containerName = resolution.definition.registrationName
```
Comment (adrien2p) [inline packages/modules-sdk/src/medusa-module.ts:53]: suggestion: ahah I come back again, but can we rename `containerName` to just `registrationName`

File: packages/stock-location/src/services/stock-location.ts [hunk 8]
```diff
@@ -167,7 +160,7 @@ export default class StockLocationService extends TransactionBaseService {
     stockLocationId: string,
     updateData: UpdateStockLocationInput
   ): Promise<StockLocation> {
-    return await this.atomicPhase_(async (manager) => {
+    return await this.manager_.transaction(async (manager) => {
```
Comment (adrien2p) [inline packages/stock-location/src/services/stock-location.ts:163]: todo: Here we have a new manager for the transaction. But all the other methods called are not aware of that manager and therefore will run outside of the transaction. (retrieve, updateAddress) should those methods accept a `context` argument that optionally include a manager?

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts:163]: Same everywhere in the service

**Review by adrien2p — CHANGES_REQUESTED** [review]
first round, it looks very promising 💪 nice work

File: packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts (hunk 4 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/admin/products/transaction/create-product-variant.ts:97]: Inventory service is indeed optional, and the functions using it are only called if it is present.
`const strategy = inventoryService ? createVariantStrategyWithInventory : createSimpleVariantStrategy`
So this is the expected behavior, but TS doesn't recognize that these functions are called when the variable has value.

File: packages/stock-location/src/services/stock-location.ts (hunk 8 above)
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:163]: Here the module is isolated, and now even shared modules won't share DB transactions with the core or other modules.
So I don't think it would change much here. But maybe it worth extending `TransactionBaseService` and if someone is using the module programatically, they could apply the withTransaction to call the methods. And here we revert it to use `atomicPhase_`. wdyt?

Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:163]: But should we start to use that way, or continue to use `.withTransaction`?
I'm honestly not a big fan of sharing a context to every single method we call, but if we stick to the withTransaction method, we could just extend `TransactionBaseService`.

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts:163]: Let’s go with transactionBaseService this time around. Will see in the future with the transaction management how we plan to do. But the choosen path was to always have a context argument in which you can have things like manager, loggedInUser and stuff like that and that way it is explicit and does not depend on the class anymore

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts:163]: If we go with that, it means that the service must be called with withTransaction() without passing the manager, to get a new instance and avoid the shared manager issue.

File: packages/inventory/package.json (hunk 2 above)
Comment (srindom) [reply packages/inventory/package.json:26]: +1 -- also, I think we should avoid core dependency in all modules.

**Review by srindom — COMMENTED** [review]
This is awesome work! 🎉 Don't have any inputs so will let @adrien2p give the approving review.

File: packages/inventory/package.json (hunk 2 above)
Comment (carlos-r-l-rodrigues) [reply packages/inventory/package.json:26]: That is temporary. Reorganizing shared functions to remove the dependency of modules on the core is a bit [bigger](https://github.com/medusajs/medusa/pull/3393) task.

File: packages/stock-location/src/services/stock-location.ts (hunk 8 above)
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts]: the core won't call any module with `withTransaction` anymore. So that will be only useful if the modules are used in custom apis or other extensions.

File: packages/inventory/src/migrations/run-migration.ts (hunk 3 above)
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/migrations/run-migration.ts:10]: I think didn't get your point. This is only connecting to run the migration.

Comment (adrien2p) [reply packages/inventory/src/migrations/run-migration.ts:10]: If a module needs to create new custom repository, which is something we should do to avoid having the query builders in utilities or services, it needs to access the dataSource. Like in the core, it has to be exposed globally. You can look what I ve done in the core database loader

File: packages/stock-location/src/services/stock-location.ts (hunk 8 above)
Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: @carlos-r-l-rodrigues have you been able to look at the last comment?

Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts]: What is you suggestion then? return the `transaction()` as it was before?

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: Without the transaction management being merged, my suggestion would look like:

```ts
async update(
    stockLocationId: string,
    updateData: UpdateStockLocationInput,
    context: { manager?: EntityManager } = {}
  ): Promise<StockLocation> {
    const update = async (manager) => {
      const context_ = { manager }
      const locationRepo = manager.getRepository(StockLocation)

      const item = await this.retrieve(stockLocationId, {}, context_)

      const { address, ...data } = updateData

      if (address) {
        if (item.address_id) {
          await this.updateAddress(item.address_id, address, context_)
        } else {
          const locAddressRepo = manager.getRepository(StockLocationAddress)
          const locAddress = locAddressRepo.create(address)
          const addressResult = await locAddressRepo.save(locAddress)
          data.address_id = addressResult.id
        }
      }

      const { metadata, ...fields } = data

      const toSave = locationRepo.merge(item, fields)
      if (metadata) {
        toSave.metadata = setMetadata(toSave, metadata)
      }

      await locationRepo.save(toSave)

      await this.eventBusService_
        .withTransaction(manager)
        .emit(StockLocationService.Events.UPDATED, {
          id: stockLocationId,
        })

      return item
    }

    if (context?.manager) {
      return await update(context.manager)
    }

    return await this.manager_.transaction(async transactionManager => {
      return await update(transactionManager)
    })
  }
```

wdyt? In that case, no need to extends TransactionBaseService, we can still work with the singleton and each subsequent call is part of the existing transaction or a new transaction that is scoped to the actual running steps

Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts]: I don't like much this shared `context` to call methods. This is not super intuitive and it doesn't change much from keeping the `withTransaction` currently used everywhere.
I will keep the `TransactionBaseService` for now and let this decision for later, as it will only affect whoever is extending the module's class.

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: The problem with extending the TransactionBaseService is that service is not called with the `withTransaction(manager)` we will have some errors `QueryRunnerAlreadyReleased`. You told me that this service will not be called with `withTransction` right? Plus, the shared context has been validated already but we can re discuss if you want. It also makes it very clear that you can or not pass a manger to be used and it remove the manager being hold by the class

Comment (olivermrbl) [reply packages/stock-location/src/services/stock-location.ts]: As far as I can understand, the issue here is, that removing the `atomicPhase` and using `manager.transaction` directly will lead to a new transaction being initiated in every single service call - inside the StockLocationService. 

We don't have a mechanism on the class to ensure that if the transaction was initiated in a previous method within the same service, it will be reused.

If we don't want to add a context param to be passed around, and we don't call the StockLocationService with `withTransaction` from within the core, we need to add a mechanism similar to the atomic phase + transaction base class, such that we reuse transaction managers inside module services.

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: The updateAdress is an example of another write that is not wrapped in the same transaction as the update when not extending the transaction base service. And when extending the transaction base service, the issue is not that. It is that if the service is used concurrently, the first consumer will set the manager on the class and the second one will be executed in the transaction of the first one. Also, if someone want to update the service and add another write within the transaction, it does not seem to be very explicit that he will have to add the mechanism to support it. Wdyt?

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: Another point, the retrieve is originally outside of the transaction, which means that you can end up trying to merge the data of an item that does not exists within the open transaction, am I right?

Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts]: right, which should be using `.withTransaction` to be called.
`await this.withTrasaction(manager).updateAddress(item.address_id, address)`

I'm ok to move that to a new context parameter. I'm just not confortable making this change in one place, where everywhere else is not like this.

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: That will not work if the initiator does not call the service with withTransaction..

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts]: Discussion resolved here https://medusacorp.slack.com/archives/C04QGK1Q8C8/p1678712394648899?thread_ts=1678710918.025129&cid=C04QGK1Q8C8

File: packages/inventory/src/migrations/run-migration.ts (hunk 3 above)
Comment (adrien2p) [reply packages/inventory/src/migrations/run-migration.ts:10]: let's keep it that way for now and see later for my comment

File: packages/inventory/src/services/inventory-item.ts [hunk 9]
```diff
@@ -97,86 +111,87 @@ export default class InventoryItemService extends TransactionBaseService {
...
+      itemRepository.merge(item, data)
+      await itemRepository.save(item)
 
-        await this.eventBusService_
-          .withTransaction(manager)
-          .emit(InventoryItemService.Events.UPDATED, {
-            id: item.id,
-          })
-      }
+      await this.eventBusService_?.emit?.(InventoryItemService.Events.UPDATED, {
```
Comment (adrien2p) [inline packages/inventory/src/services/inventory-item.ts:172]: question: I suppose it is ok since it occurs at the end of the transaction block. But in case we add some more stuff after we shouldn't pass the transaction?

File: packages/medusa/src/services/__tests__/event-bus.js [hunk 10]
```diff
@@ -24,11 +24,18 @@ describe("EventBusService", () => {
...
-        manager: MockManager,
-        stagedJobRepository,
-        logger: loggerMock,
-      })
+      eventBus = new EventBusService(
+        {
+          manager: MockManager,
+          stagedJobRepository,
+          logger: loggerMock,
+        },
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/event-bus.js:32]: suggestion: Since we are using that in multiple places I usually suggest to create a fake container up there and use it where it is needed

**Review by adrien2p — CHANGES_REQUESTED** [review]

**Review by adrien2p — APPROVED** [review]
LGTM 💪

File: packages/inventory/src/services/inventory-item.ts (hunk 9 above)
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/services/inventory-item.ts:172]: here we follow the rule that modules/core don't share db transactions.
extending the eventBusService to group all messages before emitting them will probably be a needed feature in the future

File: packages/medusa/src/commands/utils/get-migrations.js [hunk 11]
```diff
@@ -249,3 +247,43 @@ export const getModuleSharedResources = (configModule, featureFlagsRouter) => {
...
+  const moduleResolutions = registerModules(configModule.modules)
+
+  for (const moduleResolution of Object.values(moduleResolutions)) {
+    if (
+      !moduleResolution.resolutionPath ||
+      moduleResolution.moduleDeclaration.scope !== "internal" ||
+      moduleResolution.moduleDeclaration.resources !== "isolated"
```
Comment (olivermrbl) [inline packages/medusa/src/commands/utils/get-migrations.js:258]: thought(non-blocking): Move to module SDK

**Review by olivermrbl — APPROVED** [review]
this is epic @carlos-r-l-rodrigues!
